MAINT: Restructure extra_special (working name) as internal library

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-19499

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR creates a new folder for the headers that provide scalar functions in the `scipy::special` namespace, moves `_lambertw.h` and its dependency `_evalpoly.h` there, and creates a private `scipy.special._get_include` to access this folder that @izaid can use when experimenting with using these functions in CuPy.

`_binom.h` has not been added yet because it is not properly header only due to its dependence on `cephes`.

#### Additional information
<!--Any additional information you think is important.-->
cc @izaid, @rgommers

izaid commented:
This looks very useful! A couple of suggestions:

- Do we want "extra_special" to be the directory, or do we want "scipy/special"?. The question is what the C++ code should look like. I'd prefer `#include <scipy/special/lambertw.h>` rather than `#include <extra_special/lambertw.h>` so it's clear what it is.
- Regardless of the name, I'd suggest putting it all inside a top "include". So this would look like "include/extra_special" or "include/scipy/special" respectively. This include directory sits in the current special module, aka where we are changing everything now.
- `_get_include()` needs to return what you currently have as "here" in that function. Essentially it should return the location of the "include" above, with "include" but not "extra_special" or "scipy/special".

steppi commented:
We'll see where things end up, but I just wanted to get the behavior working with a minimal amount of change. I see `extra_special` as just a temporary working name at this point.  I've fixed the return value of `_get_include`.

steppi commented on scipy/special/extra_special/_lambertw.h:
```diff
@@ -27,7 +27,7 @@
 
 #include "_evalpoly.h"
-#include "error.h"
+#include "../error.h"
```
@izaid, I'd missed this earlier, but I think we need this to make the include safe right?

izaid replied:
If that's where `error.h` is, then yes! But I think `error.h` should be moved into `extra_special/`

steppi commented:
OK, so my understanding of how this should go, is that `extra_special` is just being treated as another special functions library that we vendor along with `cephes`, `amos`, `specfun`', etc. This is library for all of the special function compiled kernels that SciPy doesn't take from elsewhere. Eventually, we hope this will become an external library that we include as submodule, as suggested by @rkern in the issue associated to this PR, gh-19515, and by some other commenters in gh-19404. `extra_special` is just a placeholder name for this library that I chose because I wanted to pick something so bad that it be unlikely to stick.

I no longer think it makes sense to put these functions within the `scipy::special` namespace, since ideally these namespaces should match the include structure, and there's no intention at this time for `SciPy` to officially provide this library. I think there should instead be a one level namespace which is for now the temporary working name of the library, either `extra_special`, or perhaps something else more professional sounding if having to do`#include <extra_special/lambertw.h>` is too annoying.

At the moment, the goal of this PR is just to provide the headers for internal development purposes to @izaid and anyone else who wants to get onboard with this project, so I think it makes sense to not deform the directory structure of `scipy/special` to something that might seem odd if there was no consideration of the headers being provided elsewhere.

steppi commented:
Following up on my previous comment. I've changed the nested `scipy::special` namespace to just be `extra_special`. I also cleaned up `using`s in `lambertw.h` which were outside of any namespace, and moved the helper functions in `lambert.h` to a `detail` namespace.

izaid commented:
If we're going to stick with a single namespace, can we at least just change it to be simply `special` rather than `extra_special`? That can be a reasonable placeholder name for some time.

steppi commented:
> If we're going to stick with a single namespace, can we at least just change it to be simply `special` rather than `extra_special`? That can be a reasonable placeholder name for some time.

Yes, definitely.

izaid commented on scipy/special/__init__.py:
```diff
@@ -852,3 +852,12 @@ def btdtri(*args, **kwargs):  # type: ignore [no-redef]
+
+    This function could disappear or its behavior could change at any time.
+    """
+    import os
```
Just curious, why `import os` in the method itself and not the file? Is this best practice for SciPy?

steppi replied:
It’s not good practice. I just wanted to minimize the footprint of the function; although here it’s mostly for show I guess.

rkern replied:
In `__init__.py` files, we need to minimize accidental exposure of names. We don't want `scipy.special.os` to be a thing.

steppi replied:
Yes. That’s right. Thanks @rkern. I knew I’d done it for a reason.

izaid commented:
@steppi Looks good to me! Next up - trying with CuPy.

tirthasheshpatel commented:
@steppi There seems to be a few conflicts, can you please fix those? IMO this is ready to merge otherwise!

steppi commented:
> @steppi There seems to be a few conflicts, can you please fix those? IMO this is ready to merge otherwise!

Sorry. I got codespelled. I didn't expect any of this stuff would get touched. Fixed now.

steppi commented:
There's a test failure in `special.test_trig.py`. `trig` isn't touched by this PR but is touched in https://github.com/scipy/scipy/pull/19496/files which I'd just worked on, so I had to do a double take. I checked that the failure is appearing in other PRs; not related.

tirthasheshpatel reviewed (approved):
> There's a test failure in `special.test_trig.py`. `trig` isn't touched by this PR but is touched in https://github.com/scipy/scipy/pull/19496/files which I'd just worked on, so I had to do a double take. I checked that the failure is appearing in other PRs; not related.

Thanks for clarifying @steppi!

Good to merge now. Thanks for your work on this @steppi @izaid!
